workloads/imagenet/imagenet_pytorch/workload.py

```diff
@@ -0,0 +1,235 @@
+    # has a smaller batch size
+    return {k: float(v / num_batches) for k, v in total_metrics.items()}
+
+  def _build_dataset(self,
+      data_rng: spec.RandomState,
+      split: str,
+      data_dir: str,
+      batch_size: int):
+
+    is_train = (split == "train")
+
+    normalize = transforms.Compose([
+        transforms.ToTensor(),
+        transforms.Normalize(
+            mean=[i/255 for i in self.train_mean], 
+            std=[i/255 for i in self.train_stddev])
+    ])
+    transform_config = {
+        "train": transforms.Compose([
+            transforms.RandomResizedCrop(
+                self.center_crop_size,
+                scale=self.scale_ratio_range,
+                ratio=self.aspect_ratio_range),
+            transforms.RandomHorizontalFlip(),
+            normalize]),
+        "test": transforms.Compose([
+            transforms.Resize(self.resize_size),
+            transforms.CenterCrop(self.center_crop_size),
+            normalize])
+    }
+
+    folder = {'train': 'train', 'test': 'val'}
+
+    dataset = ImageFolder(
+        os.path.join(data_dir, folder[split]),
+        transform=transform_config[split])
+
+    dataloader = torch.utils.data.DataLoader(
+        dataset,
+        batch_size=batch_size,
+        shuffle=is_train,
+        num_workers=5,
+        pin_memory=True,
+        # Note (runame): not sure why we do this?
+        drop_last=is_train,
+        # Note (runame): if we want to use fast_collate,
+        # we need to change some things
+        #collate_fn=fast_collate
+    )
+
+    if is_train:
+      dataloader = cycle(dataloader)
+
+    return dataloader
+
+  def preprocess_for_train(
+      self,
+      selected_raw_input_batch: spec.Tensor,
+      selected_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor,
+      rng: spec.RandomState) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    del rng
+    return self.preprocess_for_eval(
+        selected_raw_input_batch, selected_label_batch, None, None)
+
+  def preprocess_for_eval(
+      self,
+      raw_input_batch: spec.Tensor,
+      raw_label_batch: spec.Tensor,
+      train_mean: spec.Tensor,
+      train_stddev: spec.Tensor) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    return (raw_input_batch.float().to(DEVICE), raw_label_batch.to(DEVICE))
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
+    torch.random.manual_seed(rng[0])
+    model = resnet50()
+    if torch.cuda.device_count() > 1:
+      model = torch.nn.DataParallel(model)
+    model.to(DEVICE)
+    return model, None
+
+  def model_fn(
+      self,
+      params: spec.ParameterContainer,
+      input_batch: spec.Tensor,
+      model_state: spec.ModelAuxiliaryState,
+      mode: spec.ForwardPassMode,
+      rng: spec.RandomState,
+      update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
+    del model_state
```

Comment:
I think the `model_state` was meant to "hold auxiliary state necessary for some models, such as the current batch norm statistics" (see https://github.com/mlcommons/algorithmic-efficiency/blob/main/RULES.md#variable-update-function). Maybe, I am missing something, but shouldn't this also be available in PyTorch?

Reply:
Yeah, we could also keep a copy of the batch norm statistics in `model_state`, but I think it's a bit unnatural in PyTorch, since it is simply part of your model. If the user wants to access it, they can always do that via the model (batch norm layers have a `running_mean` and `running_var` attribute). I guess the motivation for having a separate variable for it comes from the Jax implementation.

Do you think it's best to simply extract the statistics and keep them in `model_state`? This variable won't be used in the `model_fn()` and will be passed around a bit pointlessly, but at least the user always has direct access without the need to extract the information from the model - so maybe it makes sense after all.

Reply:
I see. Currently only JAX uses the `model_state` but not PyTorch since it can basically do the same via the `model.train()` or `model.eval()` mode, right?

In that case, perhaps we should just ignore `model_sate` for PyTorch.

What do you think about this @mikerabbat, as our PyTorch expert :)
Perhaps @znado also has an opinion on this.

Reply:
yeah, this is one of the annoying diffs between the jax and pytorch abstractions, jax models don't hold any state and are purely functional (you have to supply all inputs/RNGs/BN statistics), whereas pytorch is more object oriented. So I think it should be fine to ignore this in pytorch, because it seems unnecessary.